### PR TITLE
S5773: Add checks for JavaScriptSerializer

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Partial fix for: #3282

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs [hunk 1]
```diff
@@ -165,17 +171,45 @@ private ProgramState VisitAssignmentExpression(AssignmentExpressionSyntax assign
...
+                }
+
+                var typeSymbol = semanticModel.GetTypeInfo(resolverType.Expression).Type;
+                // If the symbol cannot be determined we consider the resolver safe to avoid FPs.
+                if (typeSymbol == null)
+                {
+                    return true;
+                }
+
+                return !typeSymbol.Is(KnownType.System_Web_Script_Serialization_SimpleTypeResolver) && IsTypeSafe(typeSymbol);
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs:210]:
```suggestion
                return typeSymbol == null || (!typeSymbol.Is(KnownType.System_Web_Script_Serialization_SimpleTypeResolver) && IsTypeSafe(typeSymbol));
```
Why is this logic different than for `HasSafeBinderInitialization` ? For `HasSafeBinderInitialization` if `typeSymbol` is null, method will return false.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs [hunk 2]
```diff
@@ -188,16 +222,17 @@ private bool HasSafeBinderInitialization(InitializerExpressionSyntax initializer
...
             }
 
-            private bool IsBinderSafe(ITypeSymbol symbol) =>
-                binderValidityMap.GetOrAdd(symbol, typeSymbol =>
+            private bool IsTypeSafe(ITypeSymbol symbol) =>
+                symbolValidityMap.GetOrAdd(symbol, typeSymbol =>
                 {
-                    var declaration = GetBindToTypeMethodDeclaration(typeSymbol);
+                    var declaration = GetBindToTypeMethodDeclaration(typeSymbol) ??
+                                      GetResolveTypeMethodDeclaration(typeSymbol);
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs:233]: Although unlikely, if we have a `JavaScriptTypeResolver` that implements `BindToType`, its the `BindToType` implementation that will be checked here instead of the `ResolveType`.
I think it's ok as the likelihood of this actually happening is very small, but it can be worth it to document as a unit test

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj [hunk 3]
```diff
@@ -40,6 +40,9 @@
     <None Include="TestCasesForRuleFailure\**\*">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Include="TestCases\RestrictDeserializedTypes.JavaScriptSerializer.cs">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
```
Comment (christophe-zurn-sonarsource) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj:45]: Why is this needed ? Isn't this already covered by 
```
    <None Include="TestCases\**\*">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
```
above ?

Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj:45]: It's not needed. Was automatically added by the IDE and I forgot to remove it.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/RestrictDeserializedTypes.cs]: I've added a fix and some test cases for this.
